### PR TITLE
mesa: fix building with llvm

### LINKED ISSUE #330057: Build failure: pkgsLLVM.mesa
### Steps To Reproduce

1. build `pkgsLLVM.mesa` (works best in the [Nixpkgs LLVM WS project](https://github.com/RossComputerGuy/nixpkgs-llvm-ws))

### Build log

https://github.com/RossComputerGuy/nixpkgs-llvm-ws/runs/27943668454
https://github.com/RossComputerGuy/nixpkgs-llvm-ws/runs/27943669465

```
x86_64-unknown-linux-gnu-ld: error: undefined symbol: std::ostream& std::ostream::_M_insert<double>(double)
>>> referenced by timer.cpp.o:(spvtools::utils::Timer::Report(char const*)) in archive /nix/store/bzl0pk4742gpjbh76r0lnk3diaydrhnr-glslang-14.2.0/lib/libSPIRV-Tools.a
>>> referenced by timer.cpp.o:(spvtools::utils::Timer::Report(char const*)) in archive /nix/store/bzl0pk4742gpjbh76r0lnk3diaydrhnr-glslang-14.2.0/lib/libSPIRV-Tools.a
>>> referenced by timer.cpp.o:(spvtools::utils::Timer::Report(char const*)) in archive /nix/store/bzl0pk4742gpjbh76r0lnk3diaydrhnr-glslang-14.2.0/lib/libSPIRV-Tools.a
>>> referenced 3 more times

x86_64-unknown-linux-gnu-ld: error: undefined symbol: std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char>>::_M_mutate(unsigned long, unsigned long, char const*, unsigned long)
>>> referenced by instruction.cpp.o:(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char>> spvtools::val::Instruction::GetOperandAs<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char>>>(unsigned long) const) in archive /nix/store/bzl0pk4742gpjbh76r0lnk3diaydrhnr-glslang-14.2.0/lib/libSPIRV-Tools.a
>>> referenced by instruction.cpp.o:(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char>> spvtools::val::Instruction::GetOperandAs<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char>>>(unsigned long) const) in archive /nix/store/bzl0pk4742gpjbh76r0lnk3diaydrhnr-glslang-14.2.0/lib/libSPIRV-Tools.a
>>> referenced by instruction.cpp.o:(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char>> spvtools::val::Instruction::GetOperandAs<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char>>>(unsigned long) const) in archive /nix/store/bzl0pk4742gpjbh76r0lnk3diaydrhnr-glslang-14.2.0/lib/libSPIRV-Tools.a
>>> referenced 131 more times

x86_64-unknown-linux-gnu-ld: error: undefined symbol: std::__detail::_Prime_rehash_policy::_M_need_rehash(unsigned long, unsigned long, unsigned long) const
>>> referenced by function.cpp.o:(std::pair<std::__detail::_Node_iterator<std::pair<unsigned int const, spvtools::val::BasicBlock>, false, false>, bool> std::_Hashtable<unsigned int, std::pair<unsigned int const, spvtools::val::BasicBlock>, std::allocator<std::pair<unsigned int const, spvtools::val::BasicBlock>>, std::__detail::_Select1st, std::equal_to<unsigned int>, std::hash<unsigned int>, std::__detail::_Mod_range_hashing, std::__detail::_Default_ranged_hash, std::__detail::_Prime_rehash_policy, std::__detail::_Hashtable_traits<false, false, true>>::_M_emplace<std::pair<unsigned int const, spvtools::val::BasicBlock>>(std::integral_consta…
>>> referenced by validation_state.cpp.o:(spvtools::val::ValidationState_t::AssignNameToId(unsigned int, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char>>)) in archive /nix/store/bzl0pk4742gpjbh76r0lnk3diaydrhnr-glslang-14.2.0/lib/libSPIRV-Tools.a
>>> referenced by function.cpp.o:(std::_Function_handler<void (spvtools::val::BasicBlock const*), spvtools::CFA<spvtools::val::BasicBlock>::TraversalRoots(std::vector<spvtools::val::BasicBlock*, std::allocator<spvtools::val::BasicBlock*>> const&, std::function<std::vector<spvtools::val::BasicBlock*, std::allocator<spvtools::val::BasicBlock*>> const* (spvtools::val::BasicBlock const*)>, std::function<std::vector<spvtools::val::BasicBlock*, std::allocator<spvtools::val::BasicBlock*>> const* (spvtools::val::BasicBlock const*)>)::'lambda'(spvtools::val::BasicBlock const*)>::_M_invoke(std::_Any_data const&, spvtools::val::BasicBlock const*&&)) in…
>>> referenced 183 more times
```

### Notify maintainers

@vcunat @primeos

### Metadata

Please run `nix-shell -p nix-info --run "nix-info -m"` and paste the result.

```console
[user@system:~]$ nix-shell -p nix-info --run "nix-info -m"
output here
```

---

Add a :+1: [reaction] to [issues you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[issues you find important]: https://github.com/NixOS/nixpkgs/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc

### PR DESCRIPTION
**Pull request by RossComputerGuy**
## Description of changes

Compliments #330065

Fixes #330057

This PR ensures meson picks the right glslang when using LLVM.

## Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [x] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: pkgs/development/libraries/mesa/default.nix [hunk 1]
```diff
@@ -171,6 +171,8 @@ in stdenv.mkDerivation {
   # Needed to discover llvm-config for cross
   preConfigure = ''
     PATH=${llvmPackages.libllvm.dev}/bin:$PATH
+  '' + lib.optionalString (stdenv.targetPlatform.useLLVM or false) ''
+    export PKG_CONFIG_PATH_FOR_BUILD="${glslang}/lib/pkgconfig:$PKG_CONFIG_PATH_FOR_BUILD"
```
Comment (Artturin) [inline pkgs/development/libraries/mesa/default.nix:175]:
```suggestion
    export PKG_CONFIG_PATH_FOR_BUILD="${buildPackages.glslang}/lib/pkgconfig:$PKG_CONFIG_PATH_FOR_BUILD"
```

Otherwise it'll be wrong for cross



Why is this necessary when `depsBuildBuild` has `pkg-config` and `nativeBuildInputs` has `glslang` already.

Comment (K900) [reply pkgs/development/libraries/mesa/default.nix:175]: I think the better question is why is our glslang propagating spirv-tools.

File: pkgs/development/libraries/mesa/default.nix [hunk 2]
```diff
@@ -392,4 +394,6 @@ in stdenv.mkDerivation {
       substitutions.drivers = mesa.drivers;
     } ./llvmpipe-hook.sh;
   };
-}
+} // lib.optionalAttrs (stdenv.cc.bintools.isLLVM && lib.versionAtLeast stdenv.cc.bintools.version "17") {
+  NIX_LDFLAGS = "--undefined-version";
```
Comment (K900) [inline pkgs/development/libraries/mesa/default.nix:398]: Why?

Comment (RossComputerGuy) [reply pkgs/development/libraries/mesa/default.nix:398]:
```
FAILED: src/gallium/targets/pipe-loader/pipe_i915.so 
x86_64-unknown-linux-gnu-clang++  -o src/gallium/targets/pipe-loader/pipe_i915.so src/gallium/targets/pipe-loader/pipe_i915.so.p/pipe_i915.c.o -Wl,--as-needed -Wl,--no-undefined -shared -fPIC -Wl,-soname,pipe_i915.so -Wl,--start-group src/gallium/auxiliary/libgallium.a src/compiler/nir/libnir.a src/compiler/libcompiler.a src/util/libmesa_util.a src/util/libmesa_util_sse41.a src/util/blake3/libblake3.a src/c11/impl/libmesa_util_c11.a src/gallium/auxiliary/libgalliumvl.a src/gallium/auxiliary/libgalliumvlwinsys.a src/gallium/drivers/i915/libi915.a src/gallium/winsys/i915/drm/libi915drm.a src/intel/common/libintel_common.a src/intel/isl/libisl.a src/intel/isl/libisl_per_hw_ver40.a src/intel/dev/libintel_dev.a src/intel/isl/libisl_per_hw_ver50.a src/intel/isl/libisl_per_hw_ver60.a src/intel/isl/libisl_per_hw_ver70.a src/intel/isl/libisl_per_hw_ver75.a src/intel/isl/libisl_per_hw_ver80.a src/intel/isl/libisl_per_hw_ver90.a src/intel/isl/libisl_per_hw_ver110.a src/intel/isl/libisl_per_hw_ver120.a src/intel/isl/libisl_per_hw_ver125.a src/intel/isl/libisl_per_hw_ver200.a src/intel/isl/libisl_tiled_memcpy.a src/intel/isl/libisl_tiled_memcpy_sse41.a src/intel/common/libintel_clflushopt.a -Wl,--gc-sections -Wl,--build-id=sha1 -Wl,--version-script /build/mesa-24.1.2/build/src/gallium/targets/pipe-loader/pipe_i915.sym /nix/store/s7f31asyzskvi3xf1162kp8b8jr37iz1-zlib-x86_64-unknown-linux-gnu-1.3.1/lib/libz.so -pthread -lm /nix/store/sfz7b94dm2s8ij49hiwc1spiv577x43j-zstd-x86_64-unknown-linux-gnu-1.5.6/lib/libzstd.so /nix/store/gkbfv526nmvqijllacngrh8a80k0lhy2-libunwind-x86_64-unknown-linux-gnu-1.8.1/lib/libunwind.so /nix/store/2bymvkqsxrjkbz0ydx0w9hn780qy4hwv-libdrm-x86_64-unknown-linux-gnu-2.4.121/lib/libdrm.so -L/nix/store/c5b927mai3l88qm5wq8nzfi5m9dnhbb8-llvm-x86_64-unknown-linux-gnu-17.0.6-lib/lib -lLLVM-17 -lsensors /nix/store/ajai5zrqpmdw6sc90rxizlhqjyamzrid-libxcb-x86_64-unknown-linux-gnu-1.17.0/lib/libxcb-sync.so /nix/store/ajai5zrqpmdw6sc90rxizlhqjyamzrid-libxcb-x86_64-unknown-linux-gnu-1.17.0/lib/libxcb-present.so /nix/store/shls4kmkps1yfgcdn543qk6xiqlgi645-libxshmfence-x86_64-unknown-linux-gnu-1.3.2/lib/libxshmfence.so /nix/store/ajai5zrqpmdw6sc90rxizlhqjyamzrid-libxcb-x86_64-unknown-linux-gnu-1.17.0/lib/libxcb-xfixes.so /nix/store/ajai5zrqpmdw6sc90rxizlhqjyamzrid-libxcb-x86_64-unknown-linux-gnu-1.17.0/lib/libxcb-dri3.so /nix/store/2bymvkqsxrjkbz0ydx0w9hn780qy4hwv-libdrm-x86_64-unknown-linux-gnu-2.4.121/lib/libdrm_intel.so -Wl,--end-group
x86_64-unknown-linux-gnu-ld: error: version script assignment of 'global' to symbol 'driver_descriptor' failed: symbol not defined
clang++: error: linker command failed with exit code 1 (use -v to see invocation)
```

Comment (K900) [reply pkgs/development/libraries/mesa/default.nix:398]: That just seems like an upstream bug?

Comment (K900) [reply pkgs/development/libraries/mesa/default.nix:398]: https://gitlab.freedesktop.org/mesa/mesa/-/merge_requests/25551 possibly relevant?

Comment (RossComputerGuy) [reply pkgs/development/libraries/mesa/default.nix:398]: Oh I thought it wasn't an upstream bug since we've had this issue with multiple packages.

Comment (K900) [reply pkgs/development/libraries/mesa/default.nix:398]: It's basically always an upstream issue, it's trying to do things to symbols that don't exist. It's not fatal on ld.bfd by default, and lld only recently made it fatal, but it's always _wrong_.

Comment (RossComputerGuy) [reply pkgs/development/libraries/mesa/default.nix:398]: Yeah, though I know in one or maybe two cases where upstream refused to fix it.

File: pkgs/development/libraries/mesa/default.nix (hunk 1 above)
Comment (RossComputerGuy) [reply pkgs/development/libraries/mesa/default.nix:175]: Wdym by propagating? It provides it. I just don't know why it's seeming to prefer the build native one.

File: pkgs/development/libraries/mesa/default.nix (hunk 2 above)
Comment (RossComputerGuy) [reply pkgs/development/libraries/mesa/default.nix]: Alright, switched to using the patch

Comment (K900) [reply pkgs/development/libraries/mesa/default.nix]: Should probably apply the patch unconditionally, so that we can tell when it breaks.

File: pkgs/development/libraries/mesa/default.nix (hunk 1 above)
Comment (K900) [reply pkgs/development/libraries/mesa/default.nix:180]: It shouldn't do that, we have spirv-tools as a separate package.

File: pkgs/development/libraries/mesa/default.nix (hunk 2 above)
Comment (RossComputerGuy) [reply pkgs/development/libraries/mesa/default.nix]: Oh ok, I was trying to prevent mass rebuilds. I'll switch it to be unconditional.

**Comment by RossComputerGuy** [comment]
Mass pinged, closing
